enhancements/network/ocp-on-smart-nics.md

```diff
@@ -0,0 +1,276 @@
+need to be kept in sync with the software on the host PC (eg, as new
+functionality is added to OCP). The obvious answer to this problem is
+to run OpenShift on the NIC as well as on the host.
+
+The initial target of this work is the [NVIDIA BlueField-2 DPU].
+However, similar products are being worked on by other vendors which
+we will eventually want to support, so the solution should not be too
+specialized to this one product.
+
+[NVIDIA BlueField-2 DPU]: https://www.nvidia.com/en-us/networking/products/data-processing-unit/
+
+## Motivation
+
+### Goals
+
+- Extend OCP as needed to support installation on Smart NICs such as
+  the BlueField-2 where:
+
+    1. The NIC is capable of running OpenShift.
+
+    2. The NIC has an IPMI-addressable Baseboard Management Controller
+       that can be used to control and provision it independently of
+       its host.
+
+    3. The NIC implements SR-IOV in such a way that we can connect
+       pods on the host directly to an OVN bridge on the NIC.
+
+- Design an appropriate architecture for an environment containing a
+  mix of x86_64 hosts with "dumb" NICs, x86_64 hosts with Smart NICs,
+  and the ARM-based Smart NICs themselves. (As discussed below, such
+  an environment may involve one or more OCP clusters.)
+
+- Define how OCP installs and upgrades will work in such an
+  environment.
+
+This enhancement does not cover what we actually _do_ with the NICs
+after they are provisioned. The primary use case of OVN offload is
+covered by "[Smart NIC OVN Offload]". It is possible that we may also
+want to allow customers to move additional network, security, and
+monitoring functionality from the CPU to the DPU in the future.
+
+[Smart NIC OVN Offload]: https://github.com/openshift/enhancements/pull/732
+
+### Non-Goals
+
+- Support for ASIC- or FPGA-based Smart NICs.
+
+- Support for ARM-based Smart NICs where we would not be able to run
+  RHCOS on the NIC (eg, because the NIC requires proprietary drivers).
+
+## Design Details
+
+### Clusters Will Contain a Mix of Smart and "Dumb" NICs
+
+OVN Kubernetes is not currently capable of mixing offloaded and
+non-offloaded networking on the same node, so every pod on a node
+containing a Smart NIC must have offloaded networking, which, in the
+current architecture, means it must explicitly _request_ offloaded
+networking. In particular, this means that most pods created by
+OpenShift operators cannot be deployed on a node with a Smart NIC, and
+so all masters and at least a few workers will need to have "dumb"
+NICs.
+
+(While the exact architectural constraints may change in the future,
+it is likely that some customers will always want the possibility of
+only having Smart NICs on a subset of nodes.)
+
+### Overall Cluster Architecture
+
+There are several possible overall architectures:
+
+  - The Smart NICs could be nodes within the same OCP cluster as their
+    hosts. We would need to use some combination of labels and taints
```

Comment:
How/What is meant to set this labels/taints? Is this intending to rely on MCO? If so I don't think it supports explicit labels/taints a node creation via kubelet flags other than master/worker.

Reply:
I presume the taints would be added by MCO, CNO, SRIOV operator, etc that has visibility into the hardware of each node. It could also be added by ovn-kubernetes running on the NIC since it clearly knows the IP of its host.

I'm not sure the taint needs to happen at node creation time necessarily, just before untrusted (non control plane?) workloads can be scheduled on the machine.

Reply:
There are some details on [where](https://github.com/openshift/enhancements/pull/732/files#diff-9cb730b0bcd44036b71650e25bd939ec20d549835f76ff5b1e964c6d20d6e574R142-R144) the label is defined and [who](https://github.com/openshift/enhancements/pull/732/files#diff-9cb730b0bcd44036b71650e25bd939ec20d549835f76ff5b1e964c6d20d6e574R203-R207) applies the label in [this](https://github.com/openshift/enhancements/pull/732) enhancement.
In short, the label is applied to nodes in the same MCP by SR-IOV Operator and the offloading of ovn components to SmartNIC is triggered while the label is applied. The current idea is to first install the node as a regular openshift node, then customize it to a SmartNIC node.

Reply:
To customize how? Automatically by the operator based on hw discovery or we expect the user to add such annotaions?

Reply:
@romfreiman It is expected that user or cluster admin will first create a custom MCP and add SmartNIC x86 nodes to the pool,  then run the following steps to customize the nodes in the custom MCP:
-  (manual) Specify node labels in SR-IOV Operator API to match on the custom MCP
-  (auto) SR-IOV Operator will apply pre-defined label to the nodes in custom MCP (The label is used by Cluster Network Operator to deploy/remove ovn-kube daemonsets to/from the nodes in custom MCP)
-  (auto) SR-IOV Operator will apply pre-defined MachineConfigs to the nodes in custom MCP via MCO API
-  (auto) Nodes in custom MCP reboot
-  (auto) SR-IOV Operator configures SmartNIC device and expose SR-IOV devices as k8s extended resource
-  (auto) Nodes become ready and serving SR-IOV pod creation